### DISCUSSION TITLE
Trying create a camera-aligned wide line in vertex shader

### DISCUSSION THREAD

**Discussion by BStringhamVRSK** [google/filament#7035] [Q&A] [answered]
I'm currently using version **1.31.4** (upgrading to latest soon).

I am trying to write a "wide-line" vertex shader that aligns the line geometry to face the camera.  The **lines are defined in world or object space**.

The trouble I'm having is that I need to transform from world/object space into clip space, and then use the **material.clipSpaceTransform** to do the screen-space line thickness calculations.  The clipSpaceTransform is only available in `vertexDomain : device` shaders.  My experiments seem to show that, in those shaders, the `getClipFromWorldMatrix()` call only returns identity (or similar), rather than the matrix I actually need to transform my world/object space lines into clip space.

I am probably missing something obvious.  Maybe I need to also be using `getWorldOffset()`? (Depricated in newer versions)

Or, do I need to use a different approach?  I could (but would rather not) transform the lines into clip space on the CPU, which seems wasteful.  Or maybe there is a better solution.

Here's my vertex shader, which has NOT been thoroughly debugged:

```
material {
	name : Unlit_LineWidth,
	requires : [
		color,
		custom0,		// Contains position of other end point for line
		custom1			// Contains signed line half-width multiplier
	],
	vertexDomain : device,
	shadingModel : unlit,
	interpolation : smooth,
	parameters : [
		{
			type : float3,
			name : baseColor
		}
	],
}

vertex {
	void materialVertex(inout MaterialVertexInputs material) {
		float4 viewportRes = getResolution();	// x=width, y=heigth, z=1/width, w=1/height
		float4x4 clipFromWorldMat = getClipFromWorldMatrix();
		float3 otherEnd = getCustom0().xyz;
		float halfWidth = getCustom1().x;
		float4 posClip = clipFromWorldMat * getPosition();
		float4 otherClip = clipFromWorldMat * float4(otherEnd, 1.0);
		float3 lineVec = normalize(otherClip.xyz - posClip.xyz);
		float2 normal = float2(-lineVec.y, lineVec.x);
		float2 widthVec = normal * halfWidth * viewportRes.zw;
		posClip.xy += widthVec;
		gl_Position = posClip;

		material.clipSpaceTransform[3][0] = widthVec.x;	// X translate
		material.clipSpaceTransform[3][1] = widthVec.y;	// Y translate
	}
}

```

**Comment by romainguy** [comment]
`clipSpaceTransform` is a matrix *you* supply for when you are in the `device` `vertexDomain`. If your vertices are defined in world or object space then you shouldn't be using the `device` domain, and you would get a non-identity `getClipFromWorldMatrix()` (which has to return identity in the `device` domain since… well there are no world coordinates in that domain). tl;dr: don't change the `vertexDomain`.

**Reply by BStringhamVRSK** [reply]
How do I get the vertex shader to accept my calculations as final, clip-space results without using `device vertexDomain`?  If I disassemble the shader (using RenderDoc), it shows me that my screen-space calculations are getting overwritten by some processing that happens after my vertex shader code is complete.  Below you can see that the first assignment to `gl_Position` comes from my calculations, but then gets overwritten later by a second assignment to `gl_Position`.

```
void main()
{
instance_index = gl_InstanceID;
vec4 _359 = (objectUniforms.a[instance_index].worldFromModelMatrix[0] * mesh_position.x) + ((objectUniforms.a[instance_index].worldFromModelMatrix[1] * mesh_position.y) + ((objectUniforms.a[instance_index].worldFromModelMatrix[2] * mesh_position.z) + objectUniforms.a[instance_index].worldFromModelMatrix[3]));
vec2 _389 = (frameUniforms.e * mesh_position).xy;
vec2 _391 = normalize((frameUniforms.e * vec4(mesh_custom0.xyz, 1.0)).xy - _389);
gl_Position = vec4(fma(vec2(-_391.y, _391.x) * mesh_custom1.x, frameUniforms.l.zw, _389), 0.5, 1.0);
vertex_color = mesh_color;
vertex_worldPosition.x = _359.x;
vertex_worldPosition.y = _359.y;
vertex_worldPosition.z = _359.z;
gl_Position = frameUniforms.e * _359;
vertex_position = gl_Position;
gl_Position.z = dot(gl_Position.zw, frameUniforms.h);
}
```

**Reply by romainguy** [reply, marked as answer]
You can apply the inverse transform to go back to the source domain. Or you can use the `device` domain and just pass the matrix as a parameter.

**Reply by BStringhamVRSK** [reply]
Thanks for your help.  I got it working.  I guess my main confusion was that I still needed to modify the world-space position and can't specify that I am inputting world-space and outputting clip-space.  I have it figured out now.  Maybe the documentation could touch on this aspect in the VertexDomain section?
